massa-grpc/src/api.rs

```diff
@@ -0,0 +1,257 @@
+    grpc: &MassaGrpcService,
+    request: tonic::Request<grpc::GetNextBlockBestParentsRequest>,
+) -> Result<grpc::GetNextBlockBestParentsResponse, GrpcError> {
+    let inner_req = request.into_inner();
+    let parents = grpc
+        .consensus_controller
+        .get_best_parents()
+        .into_iter()
+        .map(|p| grpc::BlockParent {
+            block_id: p.0.to_string(),
+            period: p.1,
+        })
+        .collect();
+    Ok(grpc::GetNextBlockBestParentsResponse {
+        id: inner_req.id,
+        parents,
+    })
+}
+
+pub(crate) fn get_selector_draws(
+    grpc: &MassaGrpcService,
+    request: tonic::Request<grpc::GetSelectorDrawsRequest>,
+) -> Result<grpc::GetSelectorDrawsResponse, GrpcError> {
+    let inner_req = request.into_inner();
+    let id = inner_req.id;
+
+    let addresses = inner_req
+        .queries
+        .into_iter()
+        .map(|query| match query.filter {
+            Some(filter) => Address::from_str(filter.address.as_str()).map_err(|e| e.into()),
+            None => Err(GrpcError::InvalidArgument("filter is missing".to_string())),
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // get future draws from selector
+    let selection_draws = {
+        let cur_slot = match timeslots::get_current_latest_block_slot(
+            grpc.grpc_config.thread_count,
+            grpc.grpc_config.t0,
+            grpc.grpc_config.genesis_timestamp,
+        ) {
+            Ok(slot) => slot.unwrap_or_else(Slot::min),
+            Err(e) => {
+                warn!("failed to get current slot with error: {}", e);
+                Slot::min()
+            }
+        };
+
+        let slot_end = Slot::new(
+            cur_slot
+                .period
+                .saturating_add(grpc.grpc_config.draw_lookahead_period_count),
+            cur_slot.thread,
+        );
+        addresses
+            .iter()
+            .map(|addr| {
+                let (nt_block_draws, nt_endorsement_draws) = grpc
+                    .selector_controller
+                    .get_address_selections(addr, cur_slot, slot_end)
+                    .unwrap_or_default();
+
+                let mut proto_nt_block_draws = Vec::with_capacity(addresses.len());
+                let mut proto_nt_endorsement_draws = Vec::with_capacity(addresses.len());
+                let iterator = izip!(nt_block_draws.into_iter(), nt_endorsement_draws.into_iter());
+                for (next_block_draw, next_endorsement_draw) in iterator {
+                    proto_nt_block_draws.push(next_block_draw.into());
+                    proto_nt_endorsement_draws.push(next_endorsement_draw.into());
+                }
+
+                (proto_nt_block_draws, proto_nt_endorsement_draws)
+            })
+            .collect::<Vec<_>>()
+    };
+
+    // compile results
+    let mut res = Vec::with_capacity(addresses.len());
+    let iterator = izip!(addresses.into_iter(), selection_draws.into_iter());
```

Comment:
I'm not sure I understand the consequences of changing this, but it seems you have iters, you collect to vecs just to iterate again. It can probably be avoided.

Reply:
This code was in the old API. We adjusted it. More info in https://docs.rs/itertools/latest/itertools/macro.izip.html